doc/src/hexdoc_hexcasting/_templates/render.js.jinja

```diff
@@ -0,0 +1,371 @@
+    img.src = URL.createObjectURL(new Blob([img_data.buffer], { type: 'image/png' }))
+
+    img.setAttribute("data-start", start_dir);
+    img.setAttribute("data-string", pattern_str);
+    img.setAttribute("width", width);
+    img.setAttribute("height", height);
+    img.setAttribute("data-per-world", attr["data-per-world"].value);
+
+    img.setAttribute("class", "spell-viz");
+
+    img.setAttribute("style", "object-fit: contain");
+
+    img.innerHTML = pat.innerHTML;
+
+    pat.parentElement?.appendChild(img);
+    pat.remove()
+  }
+}
+
+//last_load_func allows the previous event listeners to be removed
+//that way we aren't doing unecessary work when you switch between a ton of options
+let last_load_func = null;
+
+//lazily renders all of the images with the chosen render options
+//loads all of the currently open ones and then adds event listeners to open the rest when they're opened
+function render_images(draw_options) {
+  let collapsibles = document.getElementsByClassName("details-collapsible");
+
+  let load_func = (event) => {
+    render_collapsible(draw_options, event.target);
+  }
+  for (let collapsible of collapsibles) {
+    // @ts-ignore
+    collapsible.removeEventListener("toggle", last_load_func, {once: true});
+
+    if (collapsible.hasAttribute("open")) {
+      render_collapsible(draw_options, collapsible);
+    } else {
+      collapsible.addEventListener("toggle", load_func, {once: true});
+    }
+  }
+
+  last_load_func = load_func;
+}
+
+//loads all of the animated canvas elements back
+//also adds an event listener to enable them when their menu is opened
+function load_animated() {
+  let collapsibles = document.getElementsByClassName("details-collapsible");
+
+  let load_func = (event) => {
+    const canvases = event.target.querySelectorAll("canvas");
+    canvases.forEach(initializeElem);
+  };
+
+  for (let collapsible of collapsibles) {
+    let patterns = Array.from(collapsible.getElementsByClassName("spell-viz"));
+
+    let open = collapsible.hasAttribute("open");
+
+    // @ts-ignore
+    collapsible.removeEventListener("toggle", last_load_func, {once: true});
+
+    if (!open) {
+      collapsible.addEventListener("toggle", load_func, {once:true})
+    }
+    
+    for (let pattern of patterns) {
+      let attr_to_transfer = ["data-start", "data-string", "width", "height", "data-per-world", "class"];
+      let canvas = document.createElement("canvas");
+      for (const index in attr_to_transfer) {
+        const attr = attr_to_transfer[index];
+        let value = pattern.getAttribute(attr);
+        canvas.setAttribute(attr, value ? value : "");
+      }
+
+      canvas.innerHTML = pattern.innerHTML;
+
+      collapsible.appendChild(canvas);
+
+      if (open) {
+        initializeElem(canvas);
+      }
+
+      pattern.remove()
+    }
+  }
+
+  last_load_func = load_func;
+
+}
+
+/** @type {Point} */
+const intersection_point = {
+  type: "Single",
+  marker: {
+    color: [255, 255, 255, 150],
```

Comment:
not really a fan of the translucent points... this one is debatable though
```suggestion
    color: [255, 255, 255, 255],
```

Reply:
As I said on Discord, I mainly added this for the Monocolor and Gradient renderers. I could make it per-palette, but if I'm separating it, it would probably be per-renderer. Otherwise, I could just make it non-transparent for all of them, I am not quite sure yet.